src/lx/print/c.c

```diff
@@ -371,54 +501,39 @@ print_io(FILE *f, const struct fsm_options *opt)
 		fprintf(f, "\t\t}\n");
 		fprintf(f, "\t}\n");
 		fprintf(f, "\n");
-		break;
 
-	case FSM_IO_STR:
-		/*
-		 * For FSM_IO_STR we treat '\0' as the end of input,
-		 * and so there's no need to distinguish it from EOF.
-		 * We return '\0' here to save the assignment.
-		 */
-		fprintf(f, "\tassert(lx->p != NULL);\n");
+		/* FIXME: This should distinguish between alloc failure
+		 * and EOF, but will require layers of interface changes. */
+		fprintf(f, "\tif (!lx_advance_end(lx, c)) { return EOF; }\n");
```

Comment:
`lx_advance_end` can fail if `lx->push` has an allocation failure, but `lx_getc` either returns a character or EOF. It should probably signal the error more obviously, but what do you think the interface should look like?